Specify generic constraints added to support nullable reference types in C# 8 (#1178)
dotnet/csharpstandard: standard/classes.md

    : type_parameter_constraints_clause
    | type_parameter_constraints_clauses type_parameter_constraints_clause
    ;

type_parameter_constraints_clause
    : 'where' type_parameter ':' type_parameter_constraints
    ;

type_parameter_constraints
    : primary_constraint (',' secondary_constraints)? (',' constructor_constraint)?
    | secondary_constraints (',' constructor_constraint)?
    | constructor_constraint
    ;

primary_constraint
    : class_type nullable_type_attribute?
    | 'class' nullable_type_attribute?
    | 'struct'

>     | 'notnull'

Author: This needs an update. `class_type` can be combined with `notnull`.

    ;

primary_constraint
    : class_type '?'?
    | 'class' '?'?
    | 'struct'
    | 'notnull'
    | 'unmanaged'
    ;

secondary_constraints
    : interface_type '?'?

>     | type_parameter

Reviewer: Shouldn't *secondary_constraints* allow *type_parameter* `?` as the first secondary constraint?  Seems strange to only allow the `?` in subsequent secondary constraints.  E.g.

```csharp
public class C<T, U, V>
    where T : U?, V
{
}
```

vs.

```csharp
public class C<T, U, V>
    where T : U, V?
{
}
```

>     | type_parameter '?'?

Reviewer 2: What is the intended meaning of *type_parameter* *nullable_type_attribute* as a constraint? The following text does not explain and while the current version of Roslyn accepts:

```
class Curious<S, T> where S : T?
```

It only accepts reference types for `S` & `T`.

Maybe some form of the (suggested removed) text on line 616 of types.md was meant to be here?

Author: Added a sentence to clarify.

    | secondary_constraints ',' interface_type '?'?

>     | secondary_constraints ',' type_parameter '?'?

Reviewer 2: I think we should review the grammar here. I can't add this as a suggested changed as it covers non-altered lines, so just code. The rules here pre-date the use of ANTLR, if we ANTLRize them and remove unneeded left-recursion (as we do elsewhere) we get:

```ANTLR
type_parameter_constraints
    : primary_constraint (',' secondary_constraints)? (',' constructor_constraint)?
    | secondary_constraints (',' constructor_constraint)?
    | constructor_constraint
    ;

primary_constraint
    : class_type '?'?
    | 'class' '?'?
    | 'struct'
    | 'notnull'
    | 'unmanaged'
    ;

secondary_constraint
    : interface_type '?'?
    | type_parameter '?'?
    ;

secondary_constraints
    : secondary_constraint (',' secondary_constraint)*
    ;
```

Optionally to improve readability we could replace the `'?'?` by adding a rule:

```ANTLR
nullable_type_attribute
    : '?'
    ;

type_parameter_constraints
    : primary_constraint (',' secondary_constraints)? (',' constructor_constraint)?
    | secondary_constraints (',' constructor_constraint)?
    | constructor_constraint
    ;

primary_constraint
    : class_type nullable_type_attribute?
    | 'class' nullable_type_attribute?
    | 'struct'
    | 'notnull'
    | 'unmanaged'
    ;

secondary_constraint
    : interface_type nullable_type_attribute?
    | type_parameter nullable_type_attribute?
    ;

secondary_constraints
    : secondary_constraint (',' secondary_constraint)*
    ;
```

The introduced *nullable_type_attribute* would then also need to be used in other grammar rules where the literal `'?'` is used as a type attribute and not in places where it is not, e.g. for the null conditional operations.

Author: I like the second one. When I was working through the grammar, I found `'?'?` hard to mentally parse and read through the grammar.

Reviewer 3: +1 for a rule for nullable_type_attribute.

Author: I added the second set of rules.

The list of constraints given in a `where` clause can include any of the following components, in this order: a single primary constraint, one or more secondary constraints, and the constructor constraint, `new()`.

> A primary constraint can be a class type, the ***reference type constraint*** `class`, the ***not null constraint*** `notnull`, the ***value type constraint*** `struct` or the ***unmanaged type constraint*** `unmanaged`.

Reviewer 3: ```suggestion
A primary constraint can be a class type, the ***reference type constraint*** `class`, the ***value type constraint*** `struct`, the ***not null constraint*** `notnull` or the ***unmanaged type constraint*** `unmanaged`.
```

Just to keep the order in the description the same as in the grammar.

> A primary constraint can be a class type, the ***reference type constraint*** `class`, the ***not null constraint*** `notnull`, the ***value type constraint*** `struct` or the ***unmanaged type constraint*** `unmanaged`.

Reviewer 3: Do we need to say "a class type optionally followed by a *nullable_type_attribute*"? We do for the secondary constraints, so I'd imagine we should here as well. (Or remove it from there. There's a description in a couple of paragraphs anyway.)

> A secondary constraint can be a *type_parameter* or *interface_type*, optionally followed by a *nullable_type_attribute*.

Reviewer 3: ```suggestion
A secondary constraint can be an *interface_type* or *type_parameter*, optionally followed by a *nullable_type_attribute*.
```

(Again, just for consistency.)

The reference type constraint specifies that a type argument used for the type parameter shall be a reference type. All class types, interface types, delegate types, array types, and type parameters known to be a reference type (as defined below) satisfy this constraint.

> The class type, reference type constraint, and secondary constraints can include the nullable type attribute. The nullability of the type argument need not match the nullability of the type parameter. The compiler may issue a warning if the nullability of the type parameter doesn't match the nullability of the type argument.

Reviewer 3: I suspect "type parameter" should be "constraint" here.

Reviewer 3: We're unsure whether `class?` is intended to mean "definitely nullable" or "might be nullable". Current description suggests that using `string` for a type parameter of `T : class?` could issue a warning.

There are three things users might want to specify:

- "T should be a non-nullable reference type"
- "T should be a nullable reference type"
- "T should be any reference type"

Currently we only have two constraints for those three cases, which is far from ideal.

Author: The 2nd isn't actually a constraint. To say "T should be a nullable reference type" you say "T is a non-nullable reference type" and then use `T?` consistently at all uses.

Reviewer 3: Clarification from Mads: `class?` is intended to mean "might be nullable". If you want "definitely nullable" you don't specify it as a constraint, you constrain T to be non-nullable, but use T? where you want to allow a null value.

Author: Also, on line 440, imply that `class` should be non-nullable, but the compiler accepts it and might warn.

Add an example pointing this out.

Author: Also, work out what `T?` means if `T` is the type parameter, and `string?` (or some other nullable reference type) is the type argument.

Author: Also note that `typeof(T?)` is an error if `T` is known to be a reference type?

Reviewer 3: > Also note that `typeof(T?)` is an error if `T` is known to be a reference type?

Yes, and it's an error in 8 if `T` is unconstrained. It's valid *but useless* in 9+ when `T` is unconstrained (`typeof(T)` and `typeof(T?)` always return the same Type). We might want to file a bug about this.

Author: > > Also note that `typeof(T?)` is an error if `T` is known to be a reference type?
> 
> Yes, and it's an error in 8 if `T` is unconstrained. It's valid _but useless_ in 9+ when `T` is unconstrained (`typeof(T)` and `typeof(T?)` always return the same Type). We might want to file a bug about this.

These changes are added in expressions.md, in section 12.8.12 (The `typeof` operator)

The list of constraints given in a `where` clause can include any of the following components, in this order: a single primary constraint, one or more secondary constraints, and the constructor constraint, `new()`.

A primary constraint can be a class type, the ***reference type constraint*** `class`, the ***value type constraint*** `struct`, the ***not null constraint*** `notnull` or the ***unmanaged type constraint*** `unmanaged`. The class type and the reference type constraint can include the *nullable_type_attribute*.

> A secondary constraint can be an *interface_type* or *type_parameter*, optionally followed by a *nullable_type_attribute*. The presence of the *nullable_type_attribute* indicates that the type argument is allowed to the the nullable reference type that corresponds to a non-nullable reference type that satisfies the constraint.

Reviewer 2: Doubled “the” typo – ”be the” maybe?

```suggestion
A secondary constraint can be an *interface_type* or *type_parameter*, optionally followed by a *nullable_type_attribute*. The presence of the *nullable_type_attribute* indicates that the type argument is allowed to be the nullable reference type that corresponds to a non-nullable reference type that satisfies the constraint.
```

The value type constraint specifies that a type argument used for the type parameter shall be a non-nullable value type. All non-nullable struct types, enum types, and type parameters having the value type constraint satisfy this constraint. Note that although classified as a value type, a nullable value type ([§8.3.12](types.md#8312-nullable-value-types)) does not satisfy the value type constraint. A type parameter having the value type constraint shall not also have the *constructor_constraint*, although it may be used as a type argument for another type parameter with a *constructor_constraint*.

The ***not null*** constraint specifies that a type argument used for the type parameter shall be a non-nullable value type or a non-nullable reference type. All non-nullable class types, interface types, delegate types, array types, struct types, enum types, and type parameters known to be a non-nullable value type or non-nullable reference type satisfy this constraint.

> > *Note*: The `System.Nullable<T>` type specifies the non-nullable value type constraint for `T`. Thus, recursively constructed types of the forms `T??` and `Nullable<Nullable<T>>` are prohibited. *end note*

Reviewer: This note would be easier to understand if it were above the `not null` constraint description.

The list of constraints given in a `where` clause can include any of the following components, in this order: a single primary constraint, one or more secondary constraints, and the constructor constraint, `new()`.

A primary constraint can be a class type, the ***reference type constraint*** `class`, the ***nullable reference type constraint*** `class?`, the ***not null*** constraint `notnull`, the ***value type constraint*** `struct` or the ***unmanaged type constraint*** `unmanaged`.

> A secondary constraint can be a *type_parameter* or *interface_type*, either optionally followed by `?`.

Reviewer 2: This doesn’t read well to me in this context, given the comment on 426 above I’ll offer two alternatives:

```suggestion
A secondary constraint can be a *type_parameter* or *interface_type*, either optionally nullable.
```

Or if the line 426 added rule is adopted:

```suggestion
A secondary constraint can be a *type_parameter* or *interface_type*, optionally followed by a *nullable_type_attribute*.
```

- If `T` has no *interface_type* constraints but has *type_parameter* constraints, its effective interface set is the union of the effective interface sets of its *type_parameter* constraints.
- If `T` has both *interface_type* constraints and *type_parameter* constraints, its effective interface set is the union of the set of dynamic erasures of its *interface_type* constraints and the effective interface sets of its *type_parameter* constraints.

> A type parameter is *known to be a non-nullable reference type* if it has the non-nullable reference type constraint or its effective base class is not `object` or `System.ValueType`.

Reviewer 3: I wonder whether a link to "effective base class" is useful here. It's not immediately clear to me what the relationship between nullability and effective base classes would be. (I suspect this is effectively "known to be a reference type, and assumed to be non-nullable" - I'm not suggesting a terminology change, but we should think about how we could make this clearer, potentially in a note.)

Author: Maybe reword this sentence to say "has the *class_type* constraint". That sets its "effective base class" to that type.

Reviewer 3: I think "non-nullable reference type constraint" here should be "reference type constraint without the nullable type attribute".

Author: Discussion:  These two sentences need to be reworked.  It's trying to say what an instance of `T` is in terms of reference types, and it's nullability.

Start with "it's known to be a reference type if...". Then, the next sentence is "It's known to be non-nullable reference type if it's known to be a reference type AND ..."

Author: fixed in next commit.

> A type parameter is *known to be a reference type* if it has the reference type constraint or it is known to be a non-nullable reference type.

Reviewer 3: I think "reference type constraint" should be "reference type constraint including the nullable type attribute".

Author: I believe this paragraph is redundant after making the change on line 610. I removed the paragraph.

- If `T` has no *interface_type* constraints but has *type_parameter* constraints, its effective interface set is the union of the effective interface sets of its *type_parameter* constraints.
- If `T` has both *interface_type* constraints and *type_parameter* constraints, its effective interface set is the union of the set of dynamic erasures of its *interface_type* constraints and the effective interface sets of its *type_parameter* constraints.

A type parameter is *known to be a non-nullable reference type* if it has the non-nullable reference type constraint or its effective base class is not `object` or `System.ValueType`.

> A type parameter is *known to be a reference type* if it has the non-nullable reference type constraint, reference type constraint or its effective base class is not `object` or `System.ValueType`.

Reviewer 2: Could this be simplified to:

```suggestion
A type parameter is *known to be a non-nullable reference type* if it has the non-nullable reference type constraint or its effective base class is not `object` or `System.ValueType`.

A type parameter is *known to be a reference type* if it has the reference type constraint or it is known to be a non-nullable reference type.
```

Reviewer 3: `its` => `it's` in the suggestion (line 614) - but then I like it.

Author: I think we're not to use contractions, so "it is".

Reviewer 2: Edited above suggestion as per preceeding two comments from @jskeet & @BillWagner.

> class C<V>
> {
>     public V f1;

> >     public C<V> f2 = null!;

Reviewer 2: It is unclear to me why in the original `f2` was initialised, and `f1` not, when both are set in the constructor. Now a *suppression* has been added…

I suggest a comment explaining the *suppression*, or the use of `default` (which leaves the explanation for the section on `default` to handle), or just drop the initialisation.

Reviewer 3: I think it would be better to drop the initialization.
